### ISSUE TITLE
[Bug] files changing while creating a tar

### ISSUE THREAD

**Issue by hg42** [machiav3lli/oandbackupx#414]
**Description**

If a file changes while creating the tar file, the backup fails.
This is necessary, to backup a consistent state (all files in the archive or ideally in all the archives should be from the same moment).

So, an app is usually stopped before it's backed up and continued after the backup.

However, there are some apps that are not stopped because they match one of these conditions:

* `UID < 10000` are ignored
* process name starts with: [EDIT: in a former version I overlooked that there is a space in front of "android"]
  `android.`
* process name contains:
    ```
    .providers.
    systemui
    ```
* package name is
    ```
    com.android.shell
    com.android.systemui
    com.android.externalstorage
    com.android.providers.media
    com.google.android.gms
    com.google.android.gsf
    ```

This is done to prevent dead locks (e.g. we cannot stop a system process that is providing services for OAndBackupX).

An example of an app that may change files in between is `com.android.providers.contacts`.
It writes a database while syncing (e.g. `databases/contacts2.db`).

There are even some additional files, that are removed when sync is finished (from lsof output, note the process name is crippled):
```
d.process.acore 22845    u0_a111   35u      REG               8,11   3112960    1573901 /data/data/com.android.providers.contacts/databases/contacts2.db
d.process.acore 22845    u0_a111   36u      REG               8,11         0    1572914 /data/data/com.android.providers.contacts/databases/contacts2.db-wal
d.process.acore 22845    u0_a111   38u      REG               8,11     32768    1572915 /data/data/com.android.providers.contacts/databases/contacts2.db-shm
```

I guess it also affects Calendar Storage and other system providers that are synced.

User apps probably don't have this problem, because their data isn't handled by a system process (to be proved).

**Steps To Reproduce**

1. start Contacts
2. start OAndBackupX
3. enter Contacts Storage
4. switch to Contacts and drag down the list to trigger sync
5. the backup should fail with a message about changed files
    (not sure if this only happens if the file shrinks)

**System Information**

I tested with a device using Google services, but I think it should also happen with Google-free phones, it's pretty basic functionality for a phone.

-----------------------------

**Additional Notes**

extract from my message on Telegram:

in fact `databases/contacts2.db` is in `com.android.providers.contacts`, which is running as `UID=10111 USER=u0_a111`, the process that maintains the file isn't stopped, because it's `android.process.acore` (which runs as `UID=10111 USER=u0_a111`). It is excluded from stopping because process name matches `android.`.

**Possible Solutions**

As a simple solution for these system providers, we could switch off syncing while the backup runs. This should also help other apps, in case they can run into similar problems (which I doubt currently).
Not sure if modifying sync setting is possible from within OABX.

Another solution would be to stop `android.process.acore`, from memory, this isn't a good idea because it handles many other tasks in the system. I guess it will create a deadlock.

Also, when databases are synced there are some more open files (`*.db-shm`, `*.wal`), I would like to backup them in a stable state instead in the middle of a sync (or write) operation. Not sure, if we can detect this sufficiently. So, from this POV, temporarily disabling sync seems to be attractive, too.

**Comment by hg42** [comment]
I find that `pm suspend` doesn't help much.
I just got 4 of 638 apps that had changed files.
Google, Deliveries, Messages and Syncthing.
With `pm suspend` I got 7, I tried several times.
It may be because `pm suspend` itself triggers some actions. Because of this, I also tried to wait 1 second after the suspend (may be too short).

Syncthing is no wonder, because I let it sync while creating a full backup, so it is constantly working and I guess it lets itself trigger by the file system notifier, so this may be a case which we cannot control.

My hope that `pm suspend` could prevent this didn't come true.

On the other hand `pm suspend` creates some problems with apps where unsuspend doesn't out them back in a running state. It is possible to miss alarms etc.

So, `pm suspend` doesn't seem worth the trouble.
Additionally without `pm suspend` the backup runs faster.

**Comment by hg42** [comment]
the cgroup freezer is generally interesting. I think, basically, cgroups could be used for freezing.
It's the technology used for containers, so there should be a lot of possibilities.
However, cgroups have to be managed by some entity and at least the ROMs I have access to don't do that for apps.
I see some cgroup assignments, but only for some system processes.

The main advantage would be that child processes are automatically assigned to the same cgroup.
However, if the app process is stopped, it cannot start a child process. So, we already handle that case.
If someone kills the app and restarts it, it will escape both methods. This could only be solved, if there would be some manager that would assign the app to the same cgroup on every start. I think this would need to be implemented in the ROM or may be a driver.

As a conclusion, this seems to be too much trouble, and too invasive for our purposes.

**Comment by hg42** [comment]
I think this should be fixed with #493
tarcmd works like other backup solutions (=no error message from tar, but storing the file as is).
Additionally there are two developer options to prevent changes (still not perfect):
* pauseApps
* pauseApps+pmSuspend
